### PR TITLE
Remove gas counter from VM interface

### PR DESCRIPTION
**Pull request by chfast**
This implements point 1. from https://github.com/ethereum/cpp-ethereum/issues/1814.

VM classes look better. Some problems in Executive. "OnOp" callback signature changed (optional).
All comments welcome.

### CODE REVIEW

**Comment by arkpar** [comment]
needs a rebase/merge

**Comment by chfast** [comment]
Should be fine know.

**Comment by chfast** [comment]
Rebased again. Please review. This is blocking me.

**Comment by gavofyork** [comment]
another rebase?

File: evmjit/libevmjit-cpp/JitVM.cpp [hunk 1]
```diff
@@ -18,27 +18,25 @@ namespace eth
...
-		auto&& output = m_fallbackVM->go(_ext, _onOp, _step);
-		m_gas = m_fallbackVM->gas(); // copy remaining gas, Executive expects it
-		return output;
+		cwarn << "Execution rejected by EVM JIT (gas limit: " << io_gas << "), executing with interpreter";
+		m_fallbackVM = VMFactory::create(VMKind::Interpreter);
+		return m_fallbackVM->go(io_gas, _ext, _onOp, _step);
 	}
 
-	m_data.gas 			= static_cast<decltype(m_data.gas)>(m_gas);
+	m_data.gas 			= static_cast<decltype(m_data.gas)>(io_gas);
```
Comment (LefterisJP) [inline evmjit/libevmjit-cpp/JitVM.cpp]: The positions of the `=` are in a funny place. I'd say either put them all normally (`a = b`) or align them all one under the other

Comment (chriseth) [inline evmjit/libevmjit-cpp/JitVM.cpp]: I would say "stylistic whitespace only at the beginning of a line".

**Comment by chriseth** [comment]
Looks good to me apart from c-style casts, but it's not that you would have introduced them to those files ;-)

File: mix/MixClient.cpp [hunk 2]
```diff
@@ -180,7 +180,7 @@ void MixClient::executeTransaction(Transaction const& _t, State& _state, bool _c
 		else
 			levels.resize(ext.depth);
 
-		machineStates.emplace_back(MachineState({steps, vm.curPC(), inst, newMemSize, vm.gas(),
+		machineStates.emplace_back(MachineState({steps, vm.curPC(), inst, newMemSize, (u256)gas,
```
Comment (LefterisJP) [inline mix/MixClient.cpp]: Maybe break this into each argument on its own line, according to the coding standards?

**Comment by LefterisJP** [comment]
Also looks good to me apart from the 2 comments I made above

File: evmjit/libevmjit-cpp/JitVM.cpp (hunk 1 above)
Comment (chfast) [inline evmjit/libevmjit-cpp/JitVM.cpp]: They are aligned "all one under the other". GitHub just displays tabs with size 8. Try https://github.com/ethereum/cpp-ethereum/pull/1860/files?ts=4.

Comment (gavofyork) [inline evmjit/libevmjit-cpp/JitVM.cpp]: lefteris/chriseth are right here - all stylistic whitespace at beginning of line.

**Pull request merged**